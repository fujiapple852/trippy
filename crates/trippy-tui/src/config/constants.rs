use crate::config::{
    AddressFamilyConfig, AddressMode, AsMode, DnsResolveMethodConfig, GeoIpMode, IcmpExtensionMode,
    LogFormat, LogSpanEvents, Mode,
};
use std::time::Duration;

/// The default value for `mode`.
pub const DEFAULT_MODE: Mode = Mode::Tui;

/// The default value for `dns-resolve-all`.
pub const DEFAULT_DNS_RESOLVE_ALL: bool = false;

/// The default value for `log-format`.
pub const DEFAULT_LOG_FORMAT: LogFormat = LogFormat::Pretty;

/// The default value for `log-span-events`.
pub const DEFAULT_LOG_SPAN_EVENTS: LogSpanEvents = LogSpanEvents::Off;

/// The default value for `log-filter`.
pub const DEFAULT_LOG_FILTER: &str = "trippy=debug";

/// The default value for `tui-preserve-screen`.
pub const DEFAULT_TUI_PRESERVE_SCREEN: bool = false;

/// The default value for `tui-as-mode`.
pub const DEFAULT_TUI_AS_MODE: AsMode = AsMode::Asn;

/// The default value for `tui-custom-columns`.
pub const DEFAULT_CUSTOM_COLUMNS: &str = "holsravbwdt";

/// The default value for `tui-icmp-extension-mode`.
pub const DEFAULT_TUI_ICMP_EXTENSION_MODE: IcmpExtensionMode = IcmpExtensionMode::Off;

/// The default value for `tui-geoip-mode`.
pub const DEFAULT_TUI_GEOIP_MODE: GeoIpMode = GeoIpMode::Off;

/// The default value for `tui-max-addrs`.
pub const DEFAULT_TUI_MAX_ADDRS: u8 = 0;

/// The default value for `tui-address-mode`.
pub const DEFAULT_TUI_ADDRESS_MODE: AddressMode = AddressMode::Host;

/// The default value for `tui-refresh-rate`.
pub const DEFAULT_TUI_REFRESH_RATE: Duration = Duration::from_millis(100);

/// The default value for `dns-resolve-method`.
pub const DEFAULT_DNS_RESOLVE_METHOD: DnsResolveMethodConfig = DnsResolveMethodConfig::System;

/// The default value for `addr-family`.
pub const DEFAULT_ADDR_FAMILY: AddressFamilyConfig = AddressFamilyConfig::System;

/// The default value for `dns-lookup-as-info`.
pub const DEFAULT_DNS_LOOKUP_AS_INFO: bool = false;

/// The default value for `dns-timeout`.
pub const DEFAULT_DNS_TIMEOUT: Duration = Duration::from_millis(5000);

/// The default value for `dns-ttl`.
pub const DEFAULT_DNS_TTL: Duration = Duration::from_secs(300);

/// The default value for `report-cycles`.
pub const DEFAULT_REPORT_CYCLES: usize = 10;

/// The minimum TUI refresh rate.
pub const TUI_MIN_REFRESH_RATE_MS: Duration = Duration::from_millis(50);

/// The maximum TUI refresh rate.
pub const TUI_MAX_REFRESH_RATE_MS: Duration = Duration::from_millis(1000);

/// The minimum socket read timeout.
pub const MIN_READ_TIMEOUT_MS: Duration = Duration::from_millis(10);

/// The maximum socket read timeout.
pub const MAX_READ_TIMEOUT_MS: Duration = Duration::from_millis(100);

/// The minimum grace duration.
pub const MIN_GRACE_DURATION_MS: Duration = Duration::from_millis(10);

/// The maximum grace duration.
pub const MAX_GRACE_DURATION_MS: Duration = Duration::from_millis(1000);

/// The minimum IPv4 packet size we allow.
pub const MIN_PACKET_SIZE_IPV4: u16 = 28;

/// The minimum IPv6 packet size we allow.
pub const MIN_PACKET_SIZE_IPV6: u16 = 48;

/// The maximum packet size we allow.
pub const MAX_PACKET_SIZE: u16 = 1024;
