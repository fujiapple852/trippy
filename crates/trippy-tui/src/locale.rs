use itertools::Itertools;
use serde::Deserialize;
use std::cell::RefCell;
use std::collections::HashMap;
use std::str::FromStr;
use std::sync::OnceLock;
use unic_langid::LanguageIdentifier;

const FALLBACK_LOCALE: &str = "en";

/// Set the locale for the application.
///
/// If the given locale is `None` the system locale is tried. If the system locale cannot be
/// determined then the fallback locale is used.
///
/// In all cases, the language part of the locale is used if the full locale is not supported.
pub fn set_locale(locale: Option<&str>) -> String {
    let new_locale = calculate_locale(locale, sys_locale::get_locale().as_deref());
    store_locale(&new_locale);
    new_locale
}

/// Get the available locales.
pub fn available_locales() -> Vec<&'static str> {
    data()
        .0
        .iter()
        .flat_map(|(_, v)| v.0.keys().map(AsRef::as_ref))
        .unique()
        .sorted_unstable()
        .collect::<Vec<_>>()
}

/// A macro to translate an item to the current locale.
#[macro_export]
macro_rules! t {
    ($key:expr) => {
        std::borrow::Cow::Borrowed($crate::locale::__translate($key))
    };
    ($key:expr, $($kt:ident = $kv:expr),+) => {
        {
            let string = t!($key);
            $(
                let string = string.replace(concat!("%{", stringify!($kt), "}"), &$kv.to_string());
            )+
            string
        }
    };
    ($key:expr, $($kt:literal => $kv:expr),+) => {
        {
            let string = t!($key);
            $(
                let string = string.replace(concat!("%{", $kt, "}"), &$kv.to_string());
            )+
            string
        }
    };
}

/// Translate an item to the current locale.
///
/// This function is public as it is used by the `t!` macro, however is not considered part of the
/// public interface.
#[doc(hidden)]
pub fn __translate(item: &str) -> &str {
    let locale = CURRENT_LOCALE.with(Clone::clone);
    let binding = locale.borrow();
    translate_locale(item, binding.as_str())
}

/// Translate an item to a specific locale.
///
/// If the item does not exists, the key is returned. Otherwise, if item does not contain the
/// locale, the fallback locale is used. If the fallback locale does not exist, the key is
/// returned.
fn translate_locale<'a>(item: &'a str, locale: &str) -> &'a str {
    if let Some(key) = data().0.get(item) {
        if let Some(value) = key.0.get(locale) {
            value
        } else if let Some(value) = key.0.get(FALLBACK_LOCALE) {
            value
        } else {
            item
        }
    } else {
        item
    }
}

/// Get the locale data.
fn data() -> &'static Data {
    static DATA: OnceLock<Data> = OnceLock::new();
    DATA.get_or_init(|| {
        toml::from_str(include_str!("../locales.toml")).expect("Failed to parse locales.toml")
    })
}

/// This is a map of a item name (i.e. `title_hops`, `awaiting_data`, etc.) to the locale `Item`.
#[derive(Debug, Deserialize)]
struct Data(HashMap<String, Item>);

/// This is a map of locale keys (i.e. `en`, `zh`, etc.) to the translated value.
#[derive(Debug, Deserialize)]
struct Item(HashMap<String, String>);

/// calculate the locale to use.
fn calculate_locale(cfg_locale: Option<&str>, sys_locale: Option<&str>) -> String {
    let preferred = cfg_locale.or(sys_locale).unwrap_or(FALLBACK_LOCALE);
    let locales = available_locales();
    locales
        .contains(&preferred)
        .then(|| preferred.to_string())
        .or_else(|| {
            LanguageIdentifier::from_str(preferred).ok().and_then(|id| {
                let lang = id.language.to_string();
                id.region
                    .map(|r| format!("{lang}-{r}"))
                    .filter(|s| locales.contains(&s.as_str()))
                    .or_else(|| locales.contains(&lang.as_str()).then_some(lang))
            })
        })
        .unwrap_or_else(|| FALLBACK_LOCALE.to_string())
}

thread_local! {
    static CURRENT_LOCALE: RefCell<String> = RefCell::new(String::from(FALLBACK_LOCALE));
}

fn store_locale(new_locale: &str) {
    CURRENT_LOCALE.with(|locale| *locale.borrow_mut() = String::from(new_locale));
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_case::test_case;

    #[test_case(None, None, "en"; "no_locale")]
    #[test_case(Some("en"), None, "en"; "cfg_locale")]
    #[test_case(None, Some("en"), "en"; "sys_locale")]
    #[test_case(Some("en"), Some("en"), "en"; "both_locales")]
    #[test_case(Some("en"), Some("zh"), "en"; "both_locales_mismatch")]
    #[test_case(Some("zh"), Some("en"), "zh"; "both_locales_mismatch_reverse")]
    #[test_case(Some("en-US"), None, "en"; "cfg_locale_dash")]
    #[test_case(None, Some("en-US"), "en"; "sys_locale_dash")]
    #[test_case(Some("en-US"), Some("en-US"), "en"; "both_locales_dash")]
    #[test_case(Some("en-US"), Some("zh-CN"), "en"; "both_locales_mismatch_dash")]
    #[test_case(Some("zh-CN"), Some("en-US"), "zh"; "both_locales_mismatch_reverse_dash")]
    #[test_case(Some("en_US"), None, "en"; "cfg_locale_underscore")]
    #[test_case(None, Some("en_US"), "en"; "sys_locale_underscore")]
    #[test_case(Some("xx"), None, "en"; "cfg_locale_unknown")]
    #[test_case(None, Some("xx"), "en"; "sys_locale_unknown")]
    #[test_case(Some("xx"), Some("xx"), "en"; "both_locales_unknown")]
    #[test_case(Some("en-"), None, "en"; "cfg_locale_invalid_dash")]
    #[test_case(Some("en_"), None, "en"; "cfg_locale_invalid_underscore")]
    #[test_case(Some("en?"), None, "en"; "cfg_locale_invalid_accepted")]
    #[test_case(Some("zh-Hant-TW"), None, "zh-TW"; "cfg_locale_ignore_script")]
    #[test_case(None, Some("zh-Hant-TW"), "zh-TW"; "sys_locale_ignore_script")]
    #[test_case(Some("zh-Hant-TW"), Some("zh-Hant-TW"), "zh-TW"; "both_locales_ignore_script")]
    fn test_set_locale(cfg_locale: Option<&str>, sys_locale: Option<&str>, expected: &str) {
        assert_eq!(calculate_locale(cfg_locale, sys_locale), expected);
    }

    #[test]
    fn test_available_languages() {
        assert_eq!(
            available_locales(),
            vec!["de", "en", "es", "fr", "it", "pt", "ru", "sv", "tr", "zh", "zh-TW"]
        );
    }

    #[test]
    fn test_data_deserialize() {
        assert!(!data().0.is_empty());
    }

    #[test]
    fn test_translate() {
        assert_eq!(translate_locale("title_hops", "en"), "Hops");
        assert_eq!(translate_locale("title_hops", "zh"), "跳");
        assert_eq!(translate_locale("title_hops", "zh-TW"), "跳");
        assert_eq!(translate_locale("unknown_item", "en"), "unknown_item");
        assert_eq!(translate_locale("unknown_locale", "xx"), "unknown_locale");
    }

    #[test]
    fn test_translate_macro() {
        assert_eq!(t!("title_hops"), "Hops");
        assert_eq!(t!("awaiting_data"), "Awaiting data...");
        assert_eq!(t!("unknown_item"), "unknown_item");
    }

    #[test]
    fn test_zh_tw_translations() {
        // Test key Traditional Chinese translations
        assert_eq!(translate_locale("auto", "zh-TW"), "自動");
        assert_eq!(translate_locale("status_failed", "zh-TW"), "失敗");
        assert_eq!(translate_locale("status_running", "zh-TW"), "執行中");
        assert_eq!(translate_locale("title_settings", "zh-TW"), "設定");
        assert_eq!(translate_locale("help_tagline", "zh-TW"), "網路診斷工具");
        assert_eq!(translate_locale("column_loss_pct", "zh-TW"), "封包遺失率");
        assert_eq!(translate_locale("rtt", "zh-TW"), "往返時間");
    }
}
