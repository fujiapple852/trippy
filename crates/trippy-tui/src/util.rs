#[cfg(test)]
pub fn insta<F: FnOnce()>(name: &str, f: F) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_suffix(name.replace(' ', "_"));
    settings.set_snapshot_path("../tests/resources/snapshots");
    settings.set_omit_expression(true);
    settings.bind(f);
}

#[cfg(test)]
pub fn remove_whitespace(mut s: String) -> String {
    s.retain(|c| !c.is_whitespace());
    s
}
