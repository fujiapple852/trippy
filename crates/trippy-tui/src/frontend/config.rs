use crate::config::{AddressMode, AsMode, GeoIpMode, TuiColumns, TuiTheme};
use crate::config::{IcmpExtensionMode, TuiBindings};
use crate::frontend::binding::Bindings;
use crate::frontend::columns::Columns;
use crate::frontend::theme::Theme;
use chrono_tz::Tz;
use std::time::Duration;

/// Tui configuration.
#[derive(Debug)]
pub struct TuiConfig {
    /// Refresh rate.
    pub refresh_rate: Duration,
    /// The maximum ttl of hops which will be masked for privacy.
    pub privacy_max_ttl: Option<u8>,
    /// Preserve screen on exit.
    pub preserve_screen: bool,
    /// How to render addresses.
    pub address_mode: AddressMode,
    /// Lookup autonomous system (AS) information.
    pub lookup_as_info: bool,
    /// How to render autonomous system (AS) data.
    pub as_mode: AsMode,
    /// How to render ICMP extensions.
    pub icmp_extension_mode: IcmpExtensionMode,
    /// How to render `GeoIp` data.
    pub geoip_mode: GeoIpMode,
    /// The maximum number of addresses to show per hop.
    pub max_addrs: Option<u8>,
    /// The Tui color theme.
    pub theme: Theme,
    /// The Tui keyboard bindings.
    pub bindings: Bindings,
    /// The columns to display in the hops table.
    pub tui_columns: Columns,
    pub geoip_mmdb_file: Option<String>,
    pub dns_resolve_all: bool,
    /// The current locale.
    pub locale: String,
    pub timezone: Option<Tz>,
}

impl TuiConfig {
    #[expect(clippy::too_many_arguments)]
    pub fn new(
        refresh_rate: Duration,
        privacy_max_ttl: Option<u8>,
        preserve_screen: bool,
        address_mode: AddressMode,
        lookup_as_info: bool,
        as_mode: AsMode,
        icmp_extension_mode: IcmpExtensionMode,
        geoip_mode: GeoIpMode,
        max_addrs: Option<u8>,
        tui_theme: TuiTheme,
        tui_bindings: &TuiBindings,
        tui_columns: &TuiColumns,
        geoip_mmdb_file: Option<String>,
        dns_resolve_all: bool,
        locale: String,
        timezone: Option<Tz>,
    ) -> Self {
        Self {
            refresh_rate,
            privacy_max_ttl,
            preserve_screen,
            address_mode,
            lookup_as_info,
            as_mode,
            icmp_extension_mode,
            geoip_mode,
            max_addrs,
            theme: Theme::from(tui_theme),
            bindings: Bindings::from(*tui_bindings),
            tui_columns: Columns::from(tui_columns.clone()),
            geoip_mmdb_file,
            dns_resolve_all,
            locale,
            timezone,
        }
    }
}
