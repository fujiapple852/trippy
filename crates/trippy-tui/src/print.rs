use crate::config::{Args, TuiCommandItem, TuiThemeItem};
use crate::locale::available_locales;
use clap::CommandFactory;
use clap_complete::Shell;
use std::process;
use strum::VariantNames;

pub fn print_tui_theme_items() {
    println!("{}", tui_theme_items());
    process::exit(0);
}

pub fn print_tui_binding_commands() {
    println!("{}", tui_binding_commands());
    process::exit(0);
}

pub fn print_config_template() {
    println!("{}", include_str!("../trippy-config-sample.toml"));
    process::exit(0);
}

pub fn print_shell_completions(shell: Shell) -> anyhow::Result<()> {
    println!("{}", shell_completions(shell)?);
    process::exit(0);
}

pub fn print_man_page() -> anyhow::Result<()> {
    println!("{}", man_page()?);
    process::exit(0);
}

pub fn print_locales() {
    println!("TUI locales: {}", available_locales().join(", "));
    process::exit(0);
}

fn tui_theme_items() -> String {
    format!(
        "TUI theme color items: {}",
        TuiThemeItem::VARIANTS.join(", ")
    )
}

fn tui_binding_commands() -> String {
    format!(
        "TUI binding commands: {}",
        TuiCommandItem::VARIANTS.join(", ")
    )
}

fn shell_completions(shell: Shell) -> anyhow::Result<String> {
    let mut cmd = Args::command();
    let name = cmd.get_name().to_string();
    let mut buffer: Vec<u8> = vec![];
    clap_complete::generate(shell, &mut cmd, name, &mut buffer);
    Ok(String::from_utf8(buffer)?)
}

fn man_page() -> anyhow::Result<String> {
    let cmd = Args::command();
    let mut buffer: Vec<u8> = vec![];
    clap_mangen::Man::new(cmd).render(&mut buffer)?;
    Ok(String::from_utf8(buffer)?)
}

#[cfg(test)]
pub mod tests {
    use super::*;
    use crate::util::{insta, remove_whitespace};
    use test_case::test_case;

    #[test_case(&tui_theme_items(), "tui theme items match"; "tui theme items match")]
    #[test_case(&tui_binding_commands(), "tui binding commands match"; "tui binding commands match")]
    #[test_case(&shell_completions(Shell::Bash).unwrap(), "generate bash shell completions"; "generate bash shell completions")]
    #[test_case(&shell_completions(Shell::Elvish).unwrap(), "generate elvish shell completions"; "generate elvish shell completions")]
    #[test_case(&shell_completions(Shell::Fish).unwrap(), "generate fish shell completions"; "generate fish shell completions")]
    #[test_case(&shell_completions(Shell::PowerShell).unwrap(), "generate powershell shell completions"; "generate powershell shell completions")]
    #[test_case(&shell_completions(Shell::Zsh).unwrap(), "generate zsh shell completions"; "generate zsh shell completions")]
    #[test_case(&man_page().unwrap(), "generate man page"; "generate man page")]
    fn test_output(actual: &str, name: &str) {
        insta(name, || {
            insta::assert_snapshot!(remove_whitespace(actual.to_string()));
        });
    }
}
