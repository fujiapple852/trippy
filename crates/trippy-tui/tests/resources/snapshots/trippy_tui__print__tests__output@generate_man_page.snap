---
source: crates/trippy-tui/src/print.rs
---
.ie\n(.g.dsAq\(aq.el.dsAq'.THtrip1"trip0.14.0-dev".SHNAMEtrip\-Anetworkdiagnostictool.SHSYNOPSIS\fBtrip\fR[\fB\-c\fR|\fB\-\-config\-file\fR][\fB\-m\fR|\fB\-\-mode\fR][\fB\-u\fR|\fB\-\-unprivileged\fR][\fB\-p\fR|\fB\-\-protocol\fR][\fB\-\-udp\fR][\fB\-\-tcp\fR][\fB\-\-icmp\fR][\fB\-F\fR|\fB\-\-addr\-family\fR][\fB\-4\fR|\fB\-\-ipv4\fR][\fB\-6\fR|\fB\-\-ipv6\fR][\fB\-P\fR|\fB\-\-target\-port\fR][\fB\-S\fR|\fB\-\-source\-port\fR][\fB\-A\fR|\fB\-\-source\-address\fR][\fB\-I\fR|\fB\-\-interface\fR][\fB\-i\fR|\fB\-\-min\-round\-duration\fR][\fB\-T\fR|\fB\-\-max\-round\-duration\fR][\fB\-g\fR|\fB\-\-grace\-duration\fR][\fB\-\-initial\-sequence\fR][\fB\-R\fR|\fB\-\-multipath\-strategy\fR][\fB\-U\fR|\fB\-\-max\-inflight\fR][\fB\-f\fR|\fB\-\-first\-ttl\fR][\fB\-t\fR|\fB\-\-max\-ttl\fR][\fB\-\-packet\-size\fR][\fB\-\-payload\-pattern\fR][\fB\-Q\fR|\fB\-\-tos\fR][\fB\-e\fR|\fB\-\-icmp\-extensions\fR][\fB\-\-read\-timeout\fR][\fB\-r\fR|\fB\-\-dns\-resolve\-method\fR][\fB\-y\fR|\fB\-\-dns\-resolve\-all\fR][\fB\-\-dns\-timeout\fR][\fB\-\-dns\-ttl\fR][\fB\-z\fR|\fB\-\-dns\-lookup\-as\-info\fR][\fB\-s\fR|\fB\-\-max\-samples\fR][\fB\-\-max\-flows\fR][\fB\-a\fR|\fB\-\-tui\-address\-mode\fR][\fB\-\-tui\-as\-mode\fR][\fB\-\-tui\-custom\-columns\fR][\fB\-\-tui\-icmp\-extension\-mode\fR][\fB\-\-tui\-geoip\-mode\fR][\fB\-M\fR|\fB\-\-tui\-max\-addrs\fR][\fB\-\-tui\-preserve\-screen\fR][\fB\-\-tui\-refresh\-rate\fR][\fB\-\-tui\-privacy\-max\-ttl\fR][\fB\-\-tui\-locale\fR][\fB\-\-tui\-timezone\fR][\fB\-\-tui\-theme\-colors\fR][\fB\-\-print\-tui\-theme\-items\fR][\fB\-\-tui\-key\-bindings\fR][\fB\-\-print\-tui\-binding\-commands\fR][\fB\-C\fR|\fB\-\-report\-cycles\fR][\fB\-G\fR|\fB\-\-geoip\-mmdb\-file\fR][\fB\-\-generate\fR][\fB\-\-generate\-man\fR][\fB\-\-print\-config\-template\fR][\fB\-\-print\-locales\fR][\fB\-\-log\-format\fR][\fB\-\-log\-filter\fR][\fB\-\-log\-span\-events\fR][\fB\-v\fR|\fB\-\-verbose\fR][\fB\-h\fR|\fB\-\-help\fR][\fB\-V\fR|\fB\-\-version\fR][\fITARGETS\fR].SHDESCRIPTIONAnetworkdiagnostictool.SHOPTIONS.TP\fB\-c\fR,\fB\-\-config\-file\fR=\fICONFIG_FILE\fRConfigfile.TP\fB\-m\fR,\fB\-\-mode\fR=\fIMODE\fROutputmode[default:tui].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2tui:DisplayinteractiveTUI.IP\(bu2stream:Displayacontinuousstreamoftracingdata.IP\(bu2pretty:GenerateaprettytexttablereportforNcycles.IP\(bu2markdown:GenerateaMarkdowntexttablereportforNcycles.IP\(bu2csv:GenerateaCSVreportforNcycles.IP\(bu2json:GenerateaJSONreportforNcycles.IP\(bu2dot:GenerateaGraphvizDOTfileforNcycles.IP\(bu2flows:DisplayallflowsforNcycles.IP\(bu2silent:DonotgenerateanytracingoutputforNcycles.RE.TP\fB\-u\fR,\fB\-\-unprivileged\fRTracewithoutrequiringelevatedprivilegesonsupportedplatforms[default:false].TP\fB\-p\fR,\fB\-\-protocol\fR=\fIPROTOCOL\fRTracingprotocol[default:icmp].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2icmp:InternetControlMessageProtocol.IP\(bu2udp:UserDatagramProtocol.IP\(bu2tcp:TransmissionControlProtocol.RE.TP\fB\-\-udp\fRTraceusingtheUDPprotocol.TP\fB\-\-tcp\fRTraceusingtheTCPprotocol.TP\fB\-\-icmp\fRTraceusingtheICMPprotocol.TP\fB\-F\fR,\fB\-\-addr\-family\fR=\fIADDR_FAMILY\fRTheaddressfamily[default:system].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2ipv4:IPv4only.IP\(bu2ipv6:IPv6only.IP\(bu2ipv6\-then\-ipv4:IPv6withafallbacktoIPv4.IP\(bu2ipv4\-then\-ipv6:IPv4withafallbacktoIPv6.IP\(bu2system:IftheOSresolverisbeingusedthenusethefirstIPaddressreturned,otherwiselookupIPv4withafallbacktoIPv6.RE.TP\fB\-4\fR,\fB\-\-ipv4\fRUseIPv4only.TP\fB\-6\fR,\fB\-\-ipv6\fRUseIPv6only.TP\fB\-P\fR,\fB\-\-target\-port\fR=\fITARGET_PORT\fRThetargetport(TCP&UDPonly)[default:80].TP\fB\-S\fR,\fB\-\-source\-port\fR=\fISOURCE_PORT\fRThesourceport(TCP&UDPonly)[default:auto].TP\fB\-A\fR,\fB\-\-source\-address\fR=\fISOURCE_ADDRESS\fRThesourceIPaddress[default:auto].TP\fB\-I\fR,\fB\-\-interface\fR=\fIINTERFACE\fRThenetworkinterface[default:auto].TP\fB\-i\fR,\fB\-\-min\-round\-duration\fR=\fIMIN_ROUND_DURATION\fRTheminimumdurationofeveryround[default:1s].TP\fB\-T\fR,\fB\-\-max\-round\-duration\fR=\fIMAX_ROUND_DURATION\fRThemaximumdurationofeveryround[default:1s].TP\fB\-g\fR,\fB\-\-grace\-duration\fR=\fIGRACE_DURATION\fRTheperiodoftimetowaitforadditionalICMPresponsesafterthetargethasresponded[default:100ms].TP\fB\-\-initial\-sequence\fR=\fIINITIAL_SEQUENCE\fRTheinitialsequencenumber[default:33434].TP\fB\-R\fR,\fB\-\-multipath\-strategy\fR=\fIMULTIPATH_STRATEGY\fRTheEqual\-costMulti\-Pathroutingstrategy(UDPonly)[default:classic].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2classic:Thesrcordestportisusedtostorethesequencenumber.IP\(bu2paris:TheUDP`checksum`fieldisusedtostorethesequencenumber.IP\(bu2dublin:TheIP`identifier`fieldisusedtostorethesequencenumber.RE.TP\fB\-U\fR,\fB\-\-max\-inflight\fR=\fIMAX_INFLIGHT\fRThemaximumnumberofin\-flightICMPechorequests[default:24].TP\fB\-f\fR,\fB\-\-first\-ttl\fR=\fIFIRST_TTL\fRTheTTLtostartfrom[default:1].TP\fB\-t\fR,\fB\-\-max\-ttl\fR=\fIMAX_TTL\fRThemaximumnumberofTTLhops[default:64].TP\fB\-\-packet\-size\fR=\fIPACKET_SIZE\fRThesizeofIPpackettosend(IPheader+ICMPheader+payload)[default:84].TP\fB\-\-payload\-pattern\fR=\fIPAYLOAD_PATTERN\fRTherepeatingpatterninthepayloadoftheICMPpacket[default:0].TP\fB\-Q\fR,\fB\-\-tos\fR=\fITOS\fRTheTOS(i.e.DSCP+ECN)IPheadervalue(IPv4only)[default:0].TP\fB\-e\fR,\fB\-\-icmp\-extensions\fRParseICMPextensions.TP\fB\-\-read\-timeout\fR=\fIREAD_TIMEOUT\fRThesocketreadtimeout[default:10ms].TP\fB\-r\fR,\fB\-\-dns\-resolve\-method\fR=\fIDNS_RESOLVE_METHOD\fRHowtoperformDNSqueries[default:system].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2system:ResolveusingtheOSresolver.IP\(bu2resolv:Resolveusingthe`/etc/resolv.conf`DNSconfiguration.IP\(bu2google:ResolveusingtheGoogle`8.8.8.8`DNSservice.IP\(bu2cloudflare:ResolveusingtheCloudflare`1.1.1.1`DNSservice.RE.TP\fB\-y\fR,\fB\-\-dns\-resolve\-all\fRTracetoallIPsresolvedfromDNSlookup[default:false].TP\fB\-\-dns\-timeout\fR=\fIDNS_TIMEOUT\fRThemaximumtimetowaittoperformDNSqueries[default:5s].TP\fB\-\-dns\-ttl\fR=\fIDNS_TTL\fRThetime\-to\-live(TTL)ofDNSentries[default:300s].TP\fB\-z\fR,\fB\-\-dns\-lookup\-as\-info\fRLookupautonomoussystem(AS)informationduringDNSqueries[default:false].TP\fB\-s\fR,\fB\-\-max\-samples\fR=\fIMAX_SAMPLES\fRThemaximumnumberofsamplestorecordperhop[default:256].TP\fB\-\-max\-flows\fR=\fIMAX_FLOWS\fRThemaximumnumberofflowstorecord[default:64].TP\fB\-a\fR,\fB\-\-tui\-address\-mode\fR=\fITUI_ADDRESS_MODE\fRHowtorenderaddresses[default:host].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2ip:ShowIPaddressonly.IP\(bu2host:Showreverse\-lookupDNShostnameonly.IP\(bu2both:ShowbothIPaddressandreverse\-lookupDNShostname.RE.TP\fB\-\-tui\-as\-mode\fR=\fITUI_AS_MODE\fRHowtorenderautonomoussystem(AS)information[default:asn].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2asn:ShowtheASN.IP\(bu2prefix:DisplaytheASprefix.IP\(bu2country\-code:Displaythecountrycode.IP\(bu2registry:Displaytheregistryname.IP\(bu2allocated:Displaytheallocateddate.IP\(bu2name:DisplaytheASname.RE.TP\fB\-\-tui\-custom\-columns\fR=\fITUI_CUSTOM_COLUMNS\fRCustomcolumnstobedisplayedintheTUIhopstable[default:holsravbwdt].TP\fB\-\-tui\-icmp\-extension\-mode\fR=\fITUI_ICMP_EXTENSION_MODE\fRHowtorenderICMPextensions[default:off].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2off:Donotshow`icmp`extensions.IP\(bu2mpls:ShowMPLSlabel(s)only.IP\(bu2full:Showfull`icmp`extensiondataforallknownextensions.IP\(bu2all:Showfull`icmp`extensiondataforallclasses.RE.TP\fB\-\-tui\-geoip\-mode\fR=\fITUI_GEOIP_MODE\fRHowtorenderGeoIpinformation[default:short].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2off:DonotdisplayGeoIpdata.IP\(bu2short:Showshortformat.IP\(bu2long:Showlongformat.IP\(bu2location:ShowlatitudeandLongitudeformat.RE.TP\fB\-M\fR,\fB\-\-tui\-max\-addrs\fR=\fITUI_MAX_ADDRS\fRThemaximumnumberofaddressestoshowperhop[default:auto].TP\fB\-\-tui\-preserve\-screen\fRPreservethescreenonexit[default:false].TP\fB\-\-tui\-refresh\-rate\fR=\fITUI_REFRESH_RATE\fRTheTUIrefreshrate[default:100ms].TP\fB\-\-tui\-privacy\-max\-ttl\fR=\fITUI_PRIVACY_MAX_TTL\fRThemaximumttlofhopswhichwillbemaskedforprivacy[default:none]Ifset,thesourceIPaddressandhostnamewillalsobehidden..TP\fB\-\-tui\-locale\fR=\fITUI_LOCALE\fRThelocaletousefortheTUI[default:auto].TP\fB\-\-tui\-timezone\fR=\fITUI_TIMEZONE\fRThetimezonetousefortheTUI[default:auto]ThetimezonemustbeavalidIANAtimezoneidentifier..TP\fB\-\-tui\-theme\-colors\fR=\fITUI_THEME_COLORS\fRTheTUIthemecolors[item=color,item=color,..].TP\fB\-\-print\-tui\-theme\-items\fRPrintallTUIthemeitemsandexit.TP\fB\-\-tui\-key\-bindings\fR=\fITUI_KEY_BINDINGS\fRTheTUIkeybindings[command=key,command=key,..].TP\fB\-\-print\-tui\-binding\-commands\fRPrintallTUIcommandsthatcanbeboundandexit.TP\fB\-C\fR,\fB\-\-report\-cycles\fR=\fIREPORT_CYCLES\fRThenumberofreportcyclestorun[default:10].TP\fB\-G\fR,\fB\-\-geoip\-mmdb\-file\fR=\fIGEOIP_MMDB_FILE\fRThesupportedMaxMindorIPinfoGeoIpmmdbfile.TP\fB\-\-generate\fR=\fIGENERATE\fRGenerateshellcompletion.br.br[\fIpossiblevalues:\fRbash,elvish,fish,powershell,zsh].TP\fB\-\-generate\-man\fRGenerateROFFmanpage.TP\fB\-\-print\-config\-template\fRPrintatemplatetomlconfigfileandexit.TP\fB\-\-print\-locales\fRPrintallavailableTUIlocalesandexit.TP\fB\-\-log\-format\fR=\fILOG_FORMAT\fRThedebuglogformat[default:pretty].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2compact:Displaylogdatainacompactformat.IP\(bu2pretty:Displaylogdatainaprettyformat.IP\(bu2json:Displaylogdatainajsonformat.IP\(bu2chrome:DisplaylogdatainChrometraceformat.RE.TP\fB\-\-log\-filter\fR=\fILOG_FILTER\fRThedebuglogfilter[default:trippy=debug].TP\fB\-\-log\-span\-events\fR=\fILOG_SPAN_EVENTS\fRThedebuglogformat[default:off].br.br\fIPossiblevalues:\fR.RS14.IP\(bu2off:Donotdisplayeventspans.IP\(bu2active:Displayenterandexiteventspans.IP\(bu2full:Displayalleventspans.RE.TP\fB\-v\fR,\fB\-\-verbose\fREnableverbosedebuglogging.TP\fB\-h\fR,\fB\-\-help\fRPrinthelp(seeasummarywith\*(Aq\-h\*(Aq).TP\fB\-V\fR,\fB\-\-version\fRPrintversion.TP[\fITARGETS\fR]AspacedelimitedlistofhostnamesandIPstotrace.SHVERSIONv0.14.0\-dev.SHAUTHORSFujiApple<fujiapple852@gmail.com>
